packages/pigeon/CONTRIBUTING.md

```diff
@@ -0,0 +1,75 @@
+# Pigeon Contributor's Guide
+
+## Description
+
+Pigeon is a code generation tool that adds type safety to Flutter’s Platform
+Channels.  This document serves as an overview of how it functions to help
+people who would like to contribute to the project.
+
+## State Diagram
+
+Pigeon generates a temporary file in its _LaunchIsolate_, the isolate that is
+spawned to run `main()`, then launches another isolate, _PigeonIsolate_, that
+uses `dart:mirrors` to parse the generated file, creating an AST, then running
+code generators with that AST.
+
+![State Diagram](./docs/pigeon_state.png)
+
+## Source Index
+
+* [ast.dart](./lib/ast.dart) - The data structure for representing the Abstract Syntax Tree.
+* [dart_generator.dart](./lib/dart_generator.dart) - The Dart code generator.
+* [java_generator.dart](./lib/java_generator.dart) - The Java code generator.
+* [objc_generator.dart](./lib/objc_generator.dart) - The Objective-C code
+  generator (header and source files).
+* [generator_tools.dart](./lib/generator_tools.dart) - Shared code between generators.
+* [pigeon_cl.dart](./lib/pigeon_cl.dart) - The top-level function executed by
+  the command line tool in [bin/][./bin].
+* [pigeon_lib.dart](./lib/pigeon_lib.dart) - The top-level function for the
+  PigeonIsolate and the AST generation code.
+* [pigeon.dart](./lib/pigeon.dart) - A file of exported modules, the intended
+  import for users of Pigeon.
+
+## Testing Overview
+
+Pigeon has 3 types of tests, you'll find them all in [run_tests.sh](./run_tests.sh).
+
+* Unit tests - These are the fastest tests that are just typical unit tests,
+  they may be generating code and checking it against a regular expression to
+  see if it's correct.  Example:
+  [dart_generator_test.dart](./test/dart_generator_test.dart)
+* Compilation tests -  These tests generate code, then attempt to compile that
+  code.  These are tests are much slower than unit tests, but not as slow as
+  integration tests.  These tests are typically run against the Pigeon files in
+  [pigeons](./pigeons).
+* Integration tests - These tests generate code, then compile the generated
+  code, then execute the generated code.  It can be thought of as unit-tests run
+  against the generated code.  Examples: [platform_tests](./platform_tests)
+
+## Generated Source Code Example
+
+This is what the generated code that the _PigeonIsolate_ executes looks like:
```

Comment:
I still don't understand what this section means :)

Reply:
Ok, reworked it again.  It won't make sense unless you've read the State Diagram section, linked to it.